### PR TITLE
test(contracts): refactor and add unit tests to the current suite

### LINKED ISSUE #927: Smart contracts unit tests
Our current test suite for the smart contract needs to be expanded to ensure that all functions are properly tested, as well as all code paths. This will make us more confident that they work as expected.

### PR DESCRIPTION
**Pull request by ctrlc03**
# Description

Refactor of the current test suite and more test cases to ensure the smart contracts work as expected.

> Note that topup tests will be added as part of tackling issue  #923 

## Related issue(s)

fix #927 

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://github.com/privacy-scaling-explorations/maci/blob/dev/CONTRIBUTING.md) and [code of conduct](https://github.com/privacy-scaling-explorations/maci/blob/dev/CODE_OF_CONDUCT.md).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: contracts/tests/MACI.test.ts [hunk 1]
```diff
@@ -46,23 +34,46 @@ describe("MACI", () => {
...
+      const stateTreeDepthTest2 = 2;
+      const maci2 = (await deployTestContracts(initialVoiceCreditBalance, stateTreeDepthTest2, signer, true))
+        .maciContract;
+      expect(await maci2.stateTreeDepth()).to.eq(stateTreeDepthTest2);
+
+      const stateTreeDepthTest3 = 12;
+      const maci3 = (await deployTestContracts(initialVoiceCreditBalance, stateTreeDepthTest3, signer, true))
+        .maciContract;
+      expect(await maci3.stateTreeDepth()).to.eq(stateTreeDepthTest3);
+    });
```
Comment (0xmad) [inline contracts/tests/MACI.test.ts:66]:
```suggestion
    it("should be possible to deploy Maci contracts with different state tree depth values", async () => {
      const checkStateTreeDepth = async (stateTreeDepthTest: number): Promise<void> =>
        const { maciContract } = await deployTestContracts(initialVoiceCreditBalance, stateTreeDepthTest, signer, true);
        expect(await maciContract.stateTreeDepth()).to.eq(stateTreeDepthTest);
      };
      
      await checkStateTreeDepth(1);
      await checkStateTreeDepth(2);
      await checkStateTreeDepth(3);
    });
```

File: contracts/tests/Poll.test.ts [hunk 2]
```diff
@@ -33,52 +34,103 @@ describe("Poll", () => {
...
+    it("should have numMessages set to 1 (blank message)", async () => {
+      const numMessages = await pollContract.numMessages();
+      expect(numMessages.toString()).to.eq("1");
+    });
   });
 
-  describe("Publish messages (vote + key-change)", () => {
+  // describe("topup", () => {
+  //   it("should allow any user with enough credits to send a topup message", async () => {});
+  // });
```
Comment (0xmad) [inline contracts/tests/Poll.test.ts:131]: It can be removed

File: contracts/tests/PollFactory.test.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import { deployPollFactory, getDefaultSigner } from "../ts";
+import { PollFactory } from "../typechain-types";
+
+import { maxValues, messageBatchSize, tallyBatchSize, treeDepths } from "./constants";
+
+describe("pollFactory", () => {
+  let pollFactory: PollFactory;
+  let signer: Signer;
+
+  const coordinatorPubKey = new Keypair().pubKey;
```
Comment (0xmad) [inline contracts/tests/PollFactory.test.ts:14]:
```suggestion
  const { pubKey: coordinatorPubKey } = new Keypair();
```

File: contracts/ts/deploy.ts [hunk 4]
```diff
@@ -219,6 +210,26 @@ export const deployContractWithLinkedLibraries = async <T extends BaseContract>(
...
+    "PollFactory",
+    await poseidonContracts.PoseidonT3Contract.getAddress(),
+    await poseidonContracts.PoseidonT4Contract.getAddress(),
+    await poseidonContracts.PoseidonT5Contract.getAddress(),
+    await poseidonContracts.PoseidonT6Contract.getAddress(),
+    signer,
+    quiet,
+  );
+  return deployContractWithLinkedLibraries(contractFactory, "PollFactory", quiet);
+};
```
Comment (0xmad) [inline contracts/ts/deploy.ts:231]:
```suggestion
export const deployPollFactory = async (signer: Signer, quiet = false): Promise<Contract> => {
  const poseidonContracts = await deployPoseidonContracts(signer, quiet);
  const [poseidonT3Contract, poseidonT4Contract, poseidonT5Contract, poseidonT6Contract] = await Promise.all([
    poseidonContracts.PoseidonT3Contract.getAddress(),
    poseidonContracts.PoseidonT4Contract.getAddress(),
    poseidonContracts.PoseidonT5Contract.getAddress(),
    poseidonContracts.PoseidonT6Contract.getAddress(),
  ]);
  const contractFactory = await linkPoseidonLibraries(
    "PollFactory",
    poseidonT3Contract,
    poseidonT4Contract,
    poseidonT5Contract,
    poseidonT6Contract,
    signer,
    quiet,
  );
  return deployContractWithLinkedLibraries(contractFactory, "PollFactory", quiet);
};
```

**Review by 0xmad — CHANGES_REQUESTED** [review]
@ctrlc03 thanks, just few suggestions.

File: contracts/tests/MACI.test.ts [hunk 5]
```diff
@@ -303,7 +203,7 @@ describe("MACI", () => {
       pollContract = new BaseContract(pollContractAddress, pollAbi, signer) as PollContract;
     });
 
-    it("coordinator should not be able to merge the signUp AccQueue", async () => {
+    it("should not allowthe coordinator to merge the signUp AccQueue", async () => {
```
Comment (0xmad) [inline contracts/tests/MACI.test.ts:206]:
```suggestion
    it("should not allow the coordinator to merge the signUp AccQueue", async () => {
```

File: contracts/tests/MACI.test.ts [hunk 6]
```diff
@@ -330,9 +237,25 @@ describe("MACI", () => {
...
+    });
+  });
+
+  describe("getPoll", () => {
+    it("should return an address for a valid id", async () => {
+      expect(await maciContract.getPoll(pollId)).to.not.eq(0);
+    });
+    it("should throw when given an invalid poll id", async () => {
+      await expect(maciContract.getPoll(5)).to.be.revertedWithCustomError(maciContract, "PollDoesNotExist").withArgs(5);
+    });
```
Comment (0xmad) [inline contracts/tests/MACI.test.ts:259]:
```suggestion
    });

    it("should throw when given an invalid poll id", async () => {
      await expect(maciContract.getPoll(5)).to.be.revertedWithCustomError(maciContract, "PollDoesNotExist").withArgs(5);
    });
```

Comment (ctrlc03) [reply contracts/tests/MACI.test.ts:259]: prettier just puts it back to how it was 😆

**Review by 0xmad — APPROVED** [review]
@ctrlc03 thanks!

**Pull request merged**